### ISSUE TITLE
user-.slice drop-in overriding user-<UID>.slice entries

### ISSUE THREAD

**Issue by NeilW** [systemd/systemd#30366]
### systemd version the issue has been seen with

253

### Used distribution

Ubuntu 23.10

### Linux kernel version used

6.5.0-14-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

If I create a `/etc/systemd/system/user-1000.slice` with

```
[Slice]
TasksMax=3%
```

the tasks limit on the `user-1000.slice` should drop to 3%. 

### Unexpected behaviour you saw

The TasksMax setting from the `user-.slice.d` drop in directory appear to be overriding the entries in what I would expect to be the more specific `user-1000.slice` file

```
$ systemctl cat user-1000.slice
# /etc/systemd/system/user-1000.slice
[Slice]
TasksMax=3%

# /usr/lib/systemd/system/user-.slice.d/10-defaults.conf
#  SPDX-License-Identifier: LGPL-2.1-or-later
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=User Slice of UID %j
Documentation=man:user@.service(5)
StopWhenUnneeded=yes

[Slice]
TasksMax=33%
```

### Steps to reproduce the problem

As above.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
$ systemctl status user-1000.slice
● user-1000.slice - User Slice of UID 1000
     Loaded: loaded (/etc/systemd/system/user-1000.slice; static)
    Drop-In: /usr/lib/systemd/system/user-.slice.d
             └─10-defaults.conf
     Active: active since Thu 2023-12-07 12:37:08 UTC; 4min 12s ago
       Docs: man:user@.service(5)
      Tasks: 7 (limit: 2240)
     Memory: 77.3M
        CPU: 2.912s
     CGroup: /user.slice/user-1000.slice
             ├─session-4.scope
             │ ├─1081 "sshd: ubuntu [priv]"
             │ ├─1127 "sshd: ubuntu@pts/0"
             │ ├─1128 -bash
             │ ├─1139 systemctl status user-1000.slice
             │ └─1140 pager
             └─user@1000.service
               └─init.scope
                 ├─791 /lib/systemd/systemd --user
                 └─792 "(sd-pam)"
```

**Comment by NeilW** [comment]
The workaround is to use `user-1000.slice.d` drop-ins with the same name, or later in the sequence. 

This may be working as designed, but is somewhat surprising and perhaps needs documenting clearly.

**Comment by NeilW** [comment]
The filename sequence is important as systemd appears to sort the drop-ins based on the filename, which again allows `user-.slice.d` drop-ins to override drop-ins from `user-1000.slice.d`

```
$ systemctl cat user-1000.slice
# /etc/systemd/system/user-1000.slice.d/05-update.conf
[Slice]
TasksMax=3%

# /usr/lib/systemd/system/user-.slice.d/10-defaults.conf
#  SPDX-License-Identifier: LGPL-2.1-or-later
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=User Slice of UID %j
Documentation=man:user@.service(5)
StopWhenUnneeded=yes

[Slice]
TasksMax=33%
```

**Comment by YHNdnzj** [comment]
In `systemd.unit(5)`, the order in which process the drop-ins is mentioned more than once:

> Drop-in files under any of these directories take precedence over unit files wherever located. Multiple drop-in files with different names are applied in lexicographic order, regardless of which of the directories they reside in.
> ...
> The usual rules apply: multiple drop-in files with different names are applied in lexicographic order, regardless of which of the directories they reside in, so a file in type.d/ applies to a unit only if there are no drop-ins or masks with that name in directories with higher precedence. See Examples.

I think this is clear enough?

**Comment by NeilW** [comment]
There's also the line "Note that equally named drop-in files further down the prefix hierarchy override those further up, i.e. foo-bar-.service.d/10-override.conf overrides foo-.service.d/10-override.conf."

I may be worth making it crystal clear that doesn't apply when they are not equally named and they are lexicographically sorted, not prefix-hierarchy sorted.

**Comment by Werkov** [comment]
The evaluated order (same as in `systemctl cat` output) is not correct, opposite the [intention](https://github.com/systemd/systemd/blob/59b13e07f229a7b8d1ef1565b64c4f019928443a/src/shared/dropin.c#L174). I can reproduce it with v254.5.

**Comment by Werkov** [comment]
> How? The line (and the whole file) is about drop-ins, while the request was about the main config.

You're right, I was wrong. There are not drop-ins competing. The result should then be achieved by creating a drop-in instread of a main file:
`/etc/systemd/system/user-1000.slice.d/10-defaults.conf` or
`/etc/systemd/system/user-1000.slice.d/20-override.conf`.

**Comment by NeilW** [comment]
The surprising element is that an existing drop-in `/usr/lib/systemd/system/user-.slice.d/10-defaults.conf` will override *both* a more specific main slice file written in `/etc` and a more specific drop-in that just happens to have a lexical sequence lower than the existing drop-in. 

None of that is necessarily wrong, but it is somewhat surprising, and it takes a while to track down because the drop in is in the installation directories.

**Comment by YHNdnzj** [comment]
Yeah, I agree. We can consolidate the documentation for our rules of config parsing.

**Comment by piperfw** [comment]
Hi @Werkov, thanks for picking up so quickly. I was just about to correct myself, apologies. 

My mistake was I had user.slice.d/ (N.B. no dash) which acts as collective limit. Changing this to user-.slice.d, the user-1001.slice.d overrides it as expected!

**Comment by piperfw** [comment]
To clarify for anyone else with similar issues: user.slice.d provides configuration for the entire user slice, whereas user-.slice.d provides for any one particular user.

**Comment by piperfw** [comment]
I'm next intending to investigate how system.slice MemoryMin interacts with user.slice MemoryMax...

**Comment by Werkov** [comment]
> Changing this to user-.slice.d, the user-1001.slice.d overrides it as intended! Is that in line with what you expect?

Yes. (Provided the drop-ins fall in the same `/etc`|`/usr`|`/run` category. IOW, `user-.slice.d` is for any user (not all), `user-NNNN.slice.d` is for selected user, more specific overrides more general.)

**Comment by piperfw** [comment]
> `user-.slice.d` is for any user (not all together), `user-NNNN.slice.d `is for selected user, more specific overrides more general.)

Actually, I don't think this is the most helpful statement: it's true if   `user-.slice.d ` and  `user-NNNN.slice.d` have identically named files, then those in the later take precendence. However, in general lexicographical order decides precdence. IOW I would say 'more specific overrides general' is only a special case.

**Comment by Werkov** [comment]
The most useful statement in my experience is: check `systemctl cat $UNIT` to see what config files are applied (in what order) to the unit ;-)

**Comment by piperfw** [comment]
Haha, to play devil's advocate you also want to `systemctl cat $PARENT` for any parents of that unit! Thanks for the help.

**Comment by piperfw** [comment]
Hi @Werkov, sorry for resurfacing this issue. I have noticed that a later CPUQuota (for example) in the user-NNNN.slice.d is only applied if it is less than the generic user-.slice.d quota. Can I check if this is expected behaviour? I include full details of my setup below.



The followingoutput was captured in one session in the order it appears. First, I check my uid, 1229790. Second, I check the daemon has been reloaded. I then cat my user slice, which reveals i. a file from user-.slice.d which sets the CPUQuota to be 200% (2 CPUs) and ii. a file fom user-1229790.slice.d, which occurs later (apologies for the spelling error), and sets CPUQuota as 400%. There is also a system defaults file (that came with the installation) that does not define CPUQuota.

Based on my preivous understanding, I expected the CPUQuota of my user to be 400%. The machine has 4 vCPUs, which I stress with the stress-ng engine. The results, shown below, is that only 50% of TOTAL CPU time is achieved, but I expected (nearly) 100% of the TOTAL (as there are 4 cores).



```shell
$ id
uid=1229790(fik64378) gid=1229790(fik64378) groups=1229790(fik64378),6050000(global-admin),6050001(docker),6080000(local-admin)
$ sudo systemctl daemon-reload
$ systemctl cat user-1229790.slice
# /etc/systemd/system/user-.slice.d/01-individual-limits.conf
[Slice]
CPUQuota=200%

# /etc/systemd/system/user-1229790.slice.d/02-invidiaul-limits.conf
[Slice]
CPUQuota=400%

# /usr/lib/systemd/system/user-.slice.d/10-defaults.conf
#  SPDX-License-Identifier: LGPL-2.1-or-later
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=User Slice of UID %j
Documentation=man:user@.service(5)
StopWhenUnneeded=yes

[Slice]
TasksMax=33%
$ stress-ng --matrix 0 -t 30s --times
stress-ng: info:  [9162] setting to a 30 secs run per stressor
stress-ng: info:  [9162] dispatching hogs: 4 matrix
stress-ng: info:  [9162] for a 30.01s run time:
stress-ng: info:  [9162]     120.04s available CPU time
stress-ng: info:  [9162]      60.02s user time   ( 50.00%)
stress-ng: info:  [9162]       0.11s system time (  0.09%)
stress-ng: info:  [9162]      60.13s total time  ( 50.09%)
stress-ng: info:  [9162] load average: 0.00 0.02 0.08
stress-ng: info:  [9162] skipped: 0
stress-ng: info:  [9162] passed: 4: matrix (4)
stress-ng: info:  [9162] failed: 0
stress-ng: info:  [9162] metrics untrustworthy: 0
stress-ng: info:  [9162] successful run completed in 30.01 secs
```

I checked the parent slice didn't have any limits set:
```
$ systemctl cat user.slice
# /usr/lib/systemd/system/user.slice
#  SPDX-License-Identifier: LGPL-2.1-or-later
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=User and Session Slice
Documentation=man:systemd.special(7)
Before=slices.target
```

Looking in /sys/fs/cgroup/user.slice/, there does not to be any entry for user-1229790:
```shell
$ ls /sys/fs/cgroup/user.slice/ | grep user
user-1003.slice
user-116.slice
$ cat /sys/fs/cgroup/user.slice/cpu.max
max 100000
```

Is the daemon supposed to have created a 1229790 cgroup files?

**Comment by Werkov** [comment]
> here does not to be any entry for user-1229790

That's weird. systemd should've created that slice. What is `systemctl status user-1229790.slice`? (Is it running?)

Actually, I smell something fishy with parsing and setting up CPUQuota, this is
on my system (systemd-257.6-1.1.x86_64):
```
hostname:~ # systemctl cat sshd|grep CPUQ
CPUQuota=200%
CPUQuota=190%
hostname:~ # systemctl show sshd|grep CPUQ
CPUQuotaPerSecUSec=1.900000s
CPUQuotaPeriodUSec=infinity
hostname:~ # cat /sys/fs/cgroup/protected.slice/sshd.service/cpu.max
max 100000

hostname:~ # systemctl edit --drop-in=72-quota sshd 
# set to 440%
hostname:~ # systemctl show sshd|grep CPUQ
CPUQuotaPerSecUSec=4.400000s
CPUQuotaPeriodUSec=infinity
hostname:~ # cat /sys/fs/cgroup/protected.slice/sshd.service/cpu.max
max 100000
hostname:~ # systemctl restart sshd # only this applied proper value
hostname:~ # cat /sys/fs/cgroup/protected.slice/sshd.service/cpu.max
440000 100000
hostname:~ # systemctl edit --drop-in=72-quota sshd # to 300%
hostname:~ # systemctl cat sshd|grep CPUQ
CPUQuota=200%
CPUQuota=300%
hostname:~ # systemctl daemon-reload # even stronger
hostname:~ # cat /sys/fs/cgroup/protected.slice/sshd.service/cpu.max
440000 100000
# new quota not applied
```

(This would be likely a different problem than the one reported against v253.)

@piperfw What's your systemd version?

**Comment by piperfw** [comment]
HI @Werkov, thanks for looking into this and apologies for the slow response. I no longer have access to that VM (it was for a workshop and was costly to run). However, looking at my files it seems at that time I was minded to `apt-cache show systemd`. I appear to have done this in two instances revealing `255.4-1ubuntu8.6` and `255.4-1ubuntu8`(?), see attached for full details. Hopefully those are accurate. Otherwise, if I find the time I will try to recreate things on a smaller test VM.

For reference I know we also had
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 24.04.2 LTS
Release:        24.04
Codename:       noble
```

[systemd-info.txt](https://github.com/user-attachments/files/21011935/systemd-info.txt)
